tools/validate/validate.cpp

```
@@ -0,0 +1,197 @@
+    }
+
+    if (tab_characters != 0) {
+        fprintf(stderr, "Validation failed: %s contains %zu tab characters.\n", filepath.c_str(), tab_characters);
+    }
+
+    if (trailing_whitespace_lines != 0) {
+        fprintf(stderr, "Validation failed: %s contains %zu lines with trailing whitespace.\n", filepath.c_str(),
+            trailing_whitespace_lines);
+    }
+}
+
+int main() {
+    static constexpr array skipped_directories{
+        ".git"sv,
+        ".vs"sv,
+        ".vscode"sv,
+        "out"sv,
+        "vcpkg"sv,
+    };
+
+    static constexpr array skipped_extensions{
+        ".dll"sv,
+        ".exe"sv,
+        ".obj"sv,
+    };
+
+    // TRANSITION, P0202R3, use constexpr is_sorted()
+    assert(is_sorted(skipped_directories.begin(), skipped_directories.end()));
+    assert(is_sorted(skipped_extensions.begin(), skipped_extensions.end()));
+
+    vector<unsigned char> buffer; // reused for performance
+
+    try {
+        for (filesystem::recursive_directory_iterator rdi{"."}, last; rdi != last; ++rdi) {
+            if (!rdi->is_regular_file()) {
+                if (rdi->is_directory()) {
+                    const string filename = rdi->path().filename().string();
+                    if (binary_search(skipped_directories.begin(), skipped_directories.end(), filename)) {
+                        rdi.disable_recursion_pending();
+                    }
+                }
+
+                continue;
+            }
+
+            const string extension = rdi->path().extension().string();
+
+            if (binary_search(skipped_extensions.begin(), skipped_extensions.end(), extension)) {
+                continue;
+            }
+
+            const string filepath = rdi->path().string();
+
+            scan_file(filepath, buffer);
+        }
+    } catch (const exception& e) {
+        fprintf(stderr, "Validation failed: Exception: %s\n", e.what());
+    } catch (...) {

```

`catch(...)` is harmful to debugging. We want to let that crash.